components/pd_client/src/client_v2.rs

```diff
@@ -0,0 +1,1412 @@
+        });
+        Ok((r, WriteFlags::default().buffer_hint(false)))
+    });
+    loop {
+        if let Err(e) = client.wait_for_ready().await {
+            warn!("failed to acquire client for heartbeat stream"; "err" => ?e);
+            let _ = GLOBAL_TIMER_HANDLE
+                .delay(std::time::Instant::now() + Duration::from_secs(REQUEST_TIMEOUT))
+                .compat()
+                .await;
+            continue;
+        }
+        let (mut hb_tx, hb_rx) = client
+            .stub()
+            .region_heartbeat_opt(client.call_option())
+            .unwrap_or_else(|e| panic!("fail to request PD {} err {:?}", "region_heartbeat", e));
+        *response_holder.lock().unwrap() = Some(hb_rx);
+        let send_res = hb_tx.send_all(&mut requests_with_flags).await;
+        if send_res.is_ok() {
+            break;
+        }
+        let _ = hb_tx.close().await;
+        reconnect.send(client.cache_version()).await.unwrap();
+    }
+}
+
+async fn bucket_loop(
+    mut client: CachedRawClient,
+    requests: mpsc::UnboundedReceiver<ReportBucketsRequest>,
+    pending: Arc<AtomicU64>,
+    mut reconnect: mpsc::UnboundedSender<u64>,
+) {
+    let mut last_report = u64::MAX;
+    let mut requests_with_flags = requests.map(|r| {
+        let last = pending.fetch_sub(1, Ordering::Relaxed);
+        // Sender will update pending at every send operation, so as long as
+        // pending task is increasing, pending count should be reported by
+        // sender.
+        if last + 10 < last_report || last == 1 {
+            PD_PENDING_BUCKETS_GAUGE.set(last as i64 - 1);
+            last_report = last;
+        }
+        if last > last_report {
+            last_report = last - 1;
+        }
+        Ok((r, WriteFlags::default()))
+    });
+    loop {
+        if let Err(e) = client.wait_for_ready().await {
+            warn!("failed to acquire client for bucket stream"; "err" => ?e);
+            let _ = GLOBAL_TIMER_HANDLE
+                .delay(std::time::Instant::now() + Duration::from_secs(REQUEST_TIMEOUT))
+                .compat()
+                .await;
+            continue;
+        }
+        let (mut bk_tx, bk_rx) = client
+            .stub()
+            .report_buckets_opt(client.call_option())
+            .unwrap_or_else(|e| panic!("fail to request PD {} err {:?}", "report_buckets", e));
+        select! {
+            send_res = bk_tx.send_all(&mut requests_with_flags).fuse() => {
+                if send_res.is_ok() {
+                    // requests are drained.
+                    break;
+                } else {
+                    warn!("region buckets stream existed: {:?}", send_res);
+                }
+            }
+            recv_res = bk_rx.fuse() => {
+                warn!("region buckets stream existed: {:?}", recv_res);
+            }
+        }
+        let _ = bk_tx.close().await;
+        reconnect.send(client.cache_version()).await.unwrap();
+    }
+}
+
+async fn tso_loop(
```

Comment:
These are not part of the client, instead the client should provide a general interface to handle streaming calls.

Reply:
Could you give me a better idea of the "interface"? `TsoRequestStream` for example is very different from heartbeat or bucket, it requires one-on-one response and uses batching.